[DOC-441] Examples for pg_cron extension (#23484)
yugabyte/yugabyte-db: docs/content/preview/explore/ysql-language-features/pg-extensions/extension-pgcron.md

- When a new pg_cron leader node is elected, no jobs are run for the first minute. Any job that were in flight on the failed node will not be retried, as their outcome is not known.

For more information on how to schedule jobs, refer to the [pg_cron documentation](https://github.com/yugabyte/yugabyte-db/blob/master/src/postgres/third-party-extensions/pg_cron/README.md).

## Examples

The following examples decribe various ways `pg_cron` can be used to automate and improve database management tasks. The tool can help maintain database performance, consistency, and reliability through scheduled jobs.

### Monitor and identify

Use [pg_stat_statements](../extension-pgstatstatements/) to capture statistics about queries, and schedule regular reports with `pg_cron` to summarize slow queries.

```sql

> SELECT * FROM pg_stat_statements ORDER BY total_time DESC LIMIT 10;

Reviewer: ```suggestion
INSERT INTO slow_queries SELECT * FROM pg_stat_statements ORDER BY total_time DESC LIMIT 10;
```

```sql
SELECT cron.schedule('weekly_check_performance', '0 4 * * 0',
$$
DO $$
BEGIN
-- Example of weekly check
IF EXISTS (SELECT 1 FROM your_table WHERE performance_degraded) THEN
RAISE NOTICE 'Performance issue detected';
END IF;
END;
$$;
$$);
```

### Automate data refresh for materialized views

You can keep materialized views up to date ensuring that they reflect recent data changes.

This example schedules a job to refresh the `your_materialized_view` every hour.

```sql
SELECT cron.schedule('refresh_materialized_view', '0 * * * *', 'REFRESH MATERIALIZED VIEW your_materialized_view');
```

### Cleanup old logs or temporary data

You can periodically clean up old logs or temporary data to free up space.

This example schedules a job to delete logs older than 30 days every day at 2AM.

```sql
SELECT cron.schedule('cleanup_old_logs', '0 2 * * *', $$
DO $$
BEGIN
DELETE FROM logs WHERE log_date < NOW() - INTERVAL '30 days';
END;
$$);
```

> ### Best practices

Reviewer: Lets make this a ## heading and move it before Examples

#### View job details and clean up records

1. You can view the status of running and recently completed job runs in the `cron.job_run_details` table using the following command:

    ```sql
    select * from cron.job_run_details order by start_time desc limit 5;
    ```

1. The records in `cron.job_run_details` are not cleaned automatically. So in scenarios when you have jobs that run every few seconds, it can be a good idea to clean up regularly, which can easily be done using `pg_cron` as follows:

    ```sql
    -- Delete old cron.job_run_details records of the current user every day at noon
    SELECT  cron.schedule('delete-job-run-details', '0 12 * * *', $$DELETE FROM cron.job_run_details WHERE end_time < now() - interval '7 days'$$);
    ```

>     If you do not want to use `cron.job_run_details` at all, then you can add `cron.log_run = off` to `postgresql.conf`.

Reviewer: Lets remove this. This does not work in yb

> #### Create a cleanup function

Reviewer: I dont think we need this. It looks very similar to the previous one.

- When a new pg_cron leader node is elected, no jobs are run for the first minute. Any job that were in flight on the failed node will not be retried, as their outcome is not known.

For more information on how to schedule jobs, refer to the [pg_cron documentation](https://github.com/yugabyte/yugabyte-db/blob/master/src/postgres/third-party-extensions/pg_cron/README.md).

## Best practices

> As a best practice, you can set up a periodic cleanup task for the `cron.job_run_details` table using the `pg_cron` extension ensuring old data doesn't accumulate and affect database performance. The `cron.job_run_details` table is part of the `pg_cron` extension in PostgreSQL. This table keeps a record of the execution details of the scheduled cron jobs. It logs information about each cron job run, including its start and end time, status, and any exit messages or errors that occurred during the execution.

Reviewer 2: ```suggestion
The `cron.job_run_details` table is part of the `pg_cron` extension in PostgreSQL. This table keeps a record of the execution details of the scheduled cron jobs. It logs information about each cron job run, including its start and end time, status, and any exit messages or errors that occurred during the execution. The records in `cron.job_run_details` are not cleaned automatically, so in scenarios where you have jobs that run every few seconds, set up a periodic cleanup task for the table using `pg_cron` to ensure old data doesn't accumulate and affect database performance.
```

### View job details and clean up records

> 1. You can view the status of running and recently completed job runs in the `cron.job_run_details` table using the following command:

Reviewer 2: ```suggestion
1. You can view the status of running and recently completed jobs in the `cron.job_run_details` table using the following command:
```

    ```sql
    select * from cron.job_run_details order by start_time desc limit 5;
    ```

> 

Author: Skip the example ? Or just the second part ? @hari90

> 

Reviewer: On second thought it is useful. Lets maybe move it out as its own tip instead of putting it in the cleanup section?

> 1. The records in `cron.job_run_details` are not cleaned automatically. So in scenarios when you have jobs that run every few seconds, it can be a good idea to clean up regularly, which can easily be done using `pg_cron` as follows:

Reviewer 2: ```suggestion
1. Create a periodoc cleaning task for the table using `pg_cron` as follows:
```

### Monitor and identify

Use [pg_stat_statements](../extension-pgstatstatements/) to capture statistics about queries, and schedule regular reports with `pg_cron` to summarize slow queries.

```sql
INSERT INTO slow_queries SELECT * FROM pg_stat_statements ORDER BY total_time DESC LIMIT 10;
```

### Deploy changes and validate

`pg_cron` can indirectly assist with query tuning in a few ways. For example, after identifying and implementing query optimizations or indexing improvements, you can monitor the impact over time using `pg_cron` to execute scripts that validate the effectiveness of your changes.

```sql
SELECT cron.schedule('weekly_check_performance', '0 4 * * 0',
$$
DO $$
BEGIN
-- Example of weekly check
IF EXISTS (SELECT 1 FROM your_table WHERE performance_degraded) THEN
RAISE NOTICE 'Performance issue detected';
END IF;
END;
$$;
$$);
```

### Automate data refresh for materialized views

You can keep materialized views up to date ensuring that they reflect recent data changes.

> This example schedules a job to refresh the `your_materialized_view` every hour.

Reviewer 2: ```suggestion
The following example schedules a job to refresh the `your_materialized_view` every hour.
```

```sql
SELECT cron.schedule('refresh_materialized_view', '0 * * * *', 'REFRESH MATERIALIZED VIEW your_materialized_view');
```

### Cleanup old logs or temporary data

You can periodically clean up old logs or temporary data to free up space.

> This example schedules a job to delete logs older than 30 days every day at 2AM.

Reviewer 2: ```suggestion
The following example schedules a job to delete logs older than 30 days every day at 2AM.
```

- When a new pg_cron leader node is elected, no jobs are run for the first minute. Any job that were in flight on the failed node will not be retried, as their outcome is not known.

For more information on how to schedule jobs, refer to the [pg_cron documentation](https://github.com/yugabyte/yugabyte-db/blob/master/src/postgres/third-party-extensions/pg_cron/README.md).

> 

Reviewer: Pg cron readme and Aurora docs have examples of this too

## Examples

The following examples decribe various ways `pg_cron` can be used to automate and improve database management tasks. The tool can help maintain database performance, consistency, and reliability through scheduled jobs.

### Monitor and identify

Use [pg_stat_statements](../extension-pgstatstatements/) to capture statistics about queries, and schedule regular reports with `pg_cron` to summarize slow queries.

```sql
SELECT * FROM pg_stat_statements ORDER BY total_time DESC LIMIT 10;
```

### Deploy changes and validate

`pg_cron` can indirectly assist with query tuning in a few ways. For example, after identifying and implementing query optimizations or indexing improvements, you can monitor the impact over time using `pg_cron` to execute scripts that validate the effectiveness of your changes.

```sql
SELECT cron.schedule('weekly_check_performance', '0 4 * * 0',
$$
DO $$
BEGIN
-- Example of weekly check
IF EXISTS (SELECT 1 FROM your_table WHERE performance_degraded) THEN
RAISE NOTICE 'Performance issue detected';
END IF;
END;
$$;
$$);
```

> ### Archive old data

Reviewer: With partman we will just move the data to a different partion. So let's just use the example of deleting old data here.
Moving can be an expensive operation.
